external-crates/move/crates/move-compiler/src/parser/syntax.rs

```diff
@@ -274,46 +323,122 @@ fn parse_comma_list_after_start<F, R>(
     start_loc: usize,
     start_token: Tok,
     end_token: Tok,
+    item_first_set: &TokenSet,
     parse_list_item: F,
     item_description: &str,
-) -> Result<Vec<R>, Box<Diagnostic>>
+) -> Vec<R>
 where
     F: Fn(&mut Context) -> Result<R, Box<Diagnostic>>,
 {
     adjust_token(context.tokens, end_token);
-    if match_token(context.tokens, end_token)? {
-        return Ok(vec![]);
-    }
     let mut v = vec![];
-    loop {
-        if context.tokens.peek() == Tok::Comma {
+    while !context.tokens.at(end_token) {
+        if context.tokens.at_set(item_first_set) {
+            match parse_list_item(context) {
+                Ok(item) => {
+                    v.push(item);
+                    adjust_token(context.tokens, end_token);
+                    if context.tokens.peek() == end_token {
+                        break;
+                    }
+                    // expect a commma - if not at stop set, consume it or advance to the next time
+                    // or end of the list
+                    if context.at_stop_set() {
+                        break;
+                    }
+                    if let Err(diag) = consume_token(context.tokens, Tok::Comma) {
+                        let at_stop_set =
+                            skip_comma_list_item(context, start_token, end_token, *diag);
+                        if at_stop_set {
+                            break;
+                        }
+                    }
+                    adjust_token(context.tokens, end_token);
+                }
+                Err(diag) => {
+                    let at_stop_set = skip_comma_list_item(context, start_token, end_token, *diag);
+                    if at_stop_set {
+                        break;
+                    }
+                }
+            }
+        } else {
             let current_loc = context.tokens.start_loc();
             let loc = make_loc(context.tokens.file_hash(), current_loc, current_loc);
-            return Err(Box::new(diag!(
+            let diag = diag!(
                 Syntax::UnexpectedToken,
                 (loc, format!("Expected {}", item_description))
-            )));
+            );
+            let at_stop_set = skip_comma_list_item(context, start_token, end_token, diag);
+            if at_stop_set {
+                break;
+            }
         }
```

Comment:
Can you simplify this by simply moving the stop_set check to the main iteraction of the loop?

```
   while !context.tokens.at(end_token) {
        if context.at_stop_set() {
            break;
        } else if context.tokens.at_set(item_first_set) {
            match parse_list_item(context) {
                Ok(item) => {
                    v.push(item);
                    adjust_token(context.tokens, end_token);
                    if context.tokens.peek() == end_token || context.at_stop_set() {
                        break;
                    } else if let Err(diag) = consume_token(context.tokens, Tok::Comma) {
                        // skip_comma_list_item(context, start_token, end_token, *diag);
                        comma_list_error_advance(context, start_token, end_token, *diag);
                    }
                    adjust_token(context.tokens, end_token);
                }
                Err(diag) => {
                    comma_list_error_advance(context, start_token, end_token, *diag);
                }
            }
            break;
        } else {
            let current_loc = context.tokens.start_loc();
            let loc = make_loc(context.tokens.file_hash(), current_loc, current_loc);
            let diag = diag!(
                Syntax::UnexpectedToken,
                (loc, format!("Expected {}", item_description))
            );
            comma_list_error_advance(context, start_token, end_token, *diag);
        }
    }
    ```